src/transformers/models/rag/modeling_tf_rag.py

```diff
@@ -0,0 +1,1828 @@
+            if argument.startswith("generator_")
+        }
+
+        # remove question_encoder, generator kwargs from kwargs
+        for key in kwargs_question_encoder.keys():
+            del kwargs["question_encoder_" + key]
+        for key in kwargs_generator.keys():
+            del kwargs["generator_" + key]
+
+        # Load and initialize the question_encoder and generator
+        # The distinction between question_encoder and generator at the model level is made
+        # by the value of the flag `is_generator` that we need to set correctly.
+        question_encoder = kwargs_question_encoder.pop("model", None)
+        if question_encoder is None:
+            assert (
+                question_encoder_pretrained_model_name_or_path is not None
+            ), "If `model` is not defined as an argument, a `question_encoder_pretrained_model_name_or_path` has to be defined"
+
+            from ..auto.modeling_tf_auto import TFAutoModel
+
+            if "config" not in kwargs_question_encoder:
+                from ..auto.configuration_auto import AutoConfig
+
+                question_encoder_config = AutoConfig.from_pretrained(question_encoder_pretrained_model_name_or_path)
+                kwargs_question_encoder["config"] = question_encoder_config
+
+            question_encoder = TFAutoModel.from_pretrained(
+                question_encoder_pretrained_model_name_or_path,
+                name="question_encoder",
+                load_weight_prefix=cls.load_weight_prefix,
+                *model_args,
+                **kwargs_question_encoder,
+            )
+
+        generator = kwargs_generator.pop("generator", None)
+        if generator is None:
+            assert (
+                generator_pretrained_model_name_or_path is not None
+            ), "If `generator_model` is not defined as an argument, a `generator_pretrained_model_name_or_path` has to be defined"
+
+            from ..auto.modeling_tf_auto import TFAutoModelForSeq2SeqLM
+
+            if "config" not in kwargs_generator:
+                from ..auto.configuration_auto import AutoConfig
+
+                generator_config = AutoConfig.from_pretrained(generator_pretrained_model_name_or_path)
+                kwargs_generator["config"] = generator_config
+
+            generator = TFAutoModelForSeq2SeqLM.from_pretrained(
+                generator_pretrained_model_name_or_path,
+                name="generator",
+                load_weight_prefix=cls.load_weight_prefix,
+                **kwargs_generator,
+            )
+
+        # instantiate config with corresponding kwargs
+        config = kwargs.get("config", None)
+        if config is None:
+            config = RagConfig.from_question_encoder_generator_configs(
+                question_encoder.config, generator.config, **kwargs
+            )
+
+        return cls(question_encoder=question_encoder, generator=generator, config=config, retriever=retriever)
+
+
+RAG_START_DOCSTRING = r"""
```

Comment:
Would be nice to indicate in the documentation that RAG can only be run in eager mode and cannot be exported in SavedModel format or even trained.

Reply:
Hi Julien, in Rag/TFRag there are two options for both forward and backward passes. One is using `input_ids` and another is using `context_input_ids`. In TFRag, I now successfully train the model with `context_input_ids` in both graph-mode and eager-mode.  I am currently working on a Colab notebook to illustrate how to do this (with some extension on TFRag code). 

So, I think I will do another PR regarding training shortly after this PR.

Reply:
This model, at least how the code looks like, is not graph compliant at all (for example it is full of `.numpy()` that cannot be run in a graph, including some conditions that are not properly setup, the usage of `.shape` that is unstable in graph mode, some keywords arguments that are not properly handled, etc.... So basically you cannot properly train it, without specifying that you want it to be run in eager mode. Hence, not possible to get a clean saved model, no proper training, almost no grappler optimizations, etc... compared to all the others. That's why I'm not very picky on the review ^^

To be sure, you can disable the eager execution with `tf.compat.v1.disable_eager_execution()` and try to run a simple inference. If an error is raised it means that your model is not graph compliant. (This is my case)

So I would like this to be specified in the documentation, this is mostly to make the users aware of this context.

Nevertheless, that's nice to be able to run it without the need of a retriever, this makes things much easier to have this model in "proper" TensorFlow. Then, if you want to have it fully graph compliant I'm fine with it but the model needs more rework. Do you want me to help with this?

Reply:
Thanks Julien. I agree that the documentation should be specified to let users know of the context. I will change documentation as you suggested. 

It would be nice to have some subsequent PRs to gradually improve on the training aspects incorporating your suggestions. 😺 
[This colab notebook ](https://colab.research.google.com/drive/1TQXg39HdUNNPbXAJ0MEdnvWW4xoZJwcS?usp=sharing) is my draft to show that the current implementation (with some udpates including points that you suggested above), a model could be trained in both graph mode and eager mode.
(Shown in two last cells of the notebook - one with `model.compile( ... , run_eagerly=False )` has 441ms/step and another with `model.compile( ... , run_eagerly=True )`  has 895ms/step so performance does have two times faster on `run_eagerly=False`. 

I am aware that this training aspect is still a draft and we need to incorporate your suggestions, organize and test properly so I would love to make an improvement on separate subsequent PRs.

Reply:
Awesome!! Good to know you have started to investigate the "next steps" for this model!! I will be more than happy to help to have a stronger TF implementation afterwards.

My main point here was to let the users know that this model will be still in an "experimental" state so that they are not surprised that it doesn't work for such or such a case 😃